internal/services/network/subnet_resource.go

```diff
@@ -230,7 +231,7 @@ func resourceSubnetCreate(d *pluginsdk.ResourceData, meta interface{}) error {
 	properties.PrivateEndpointNetworkPolicies = network.VirtualNetworkPrivateEndpointNetworkPolicies(expandSubnetPrivateLinkNetworkPolicy(privateEndpointNetworkPolicies))
 	properties.PrivateLinkServiceNetworkPolicies = network.VirtualNetworkPrivateLinkServiceNetworkPolicies(expandSubnetPrivateLinkNetworkPolicy(privateLinkServiceNetworkPolicies))
 
-	serviceEndpointsRaw := d.Get("service_endpoints").([]interface{})
+	serviceEndpointsRaw := d.Get("service_endpoints").(*pluginsdk.Set).List()
 	properties.ServiceEndpoints = expandSubnetServiceEndpoints(serviceEndpointsRaw)
```

Comment:
this'll need to be conditionally switched in 2.x/3.0 mode:

```suggestion
	if features.ThreePointOhBeta() {
		serviceEndpointsRaw := d.Get("service_endpoints").(*pluginsdk.Set).List()
		properties.ServiceEndpoints = expandSubnetServiceEndpoints(serviceEndpointsRaw)
	} else {
		serviceEndpointsRaw := d.Get("service_endpoints").([]interface{})
		properties.ServiceEndpoints = expandSubnetServiceEndpoints(serviceEndpointsRaw)
	}
```